### PR TITLE
feat: Outputs media state according to the TwicPics life cycle + refa…

### LINKED ISSUE #21: [feature request] Load callback, or passing a ref to the image?
Currently, the only way to test if an image is loaded seem to be by looking at the CSS classes of the rendered components. In a React environment it would be nice to take advantage of the virtual DOM and e.g. pass a `ref` or load/error/progress callback functions to the image component.

My use case is a slide show, where I only want to advance to the next (responsive) image once it is loaded. I would like to avoid circumventing React and using `document.querySelector`.

### PR DESCRIPTION
**Pull request by mbgspcii**
…ctoring of the placeholder management. fix #21

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
 
+### Added
+
+- Allow to track the loading status of images and videos.
```
Comment (Strift) [inline CHANGELOG.md:13]: I'd write:

Add API to listen to the loading state of images and videos.

File: src/_documentation/angular.md [hunk 2]
```diff
@@ -2,6 +2,11 @@
 // /(\b)__FRAMEWORK_NAME__(\b)/gm => "Angular"
 // /(\b)__FRAMEWORK_URL__(\b)/gm => "https://angular.io/"
 // /(\b)__INTERCOM_TERM__(\b)/gm => "angular"
+// /(\b)__TWIC_STATE_TABLE_CONTENT__(\b)/gm => "\n    - [Lifecycle](#life-cycle)"
```
Comment (Strift) [inline src/_documentation/angular.md:5]: Anchor should be `#lifecycle`

File: src/_documentation/angular.md [hunk 3]
```diff
@@ -2,6 +2,11 @@
...
+// /(\b)__TWIC_STATE_CHANGE_PROP__(\b)/gm => "\n  (stateChangeEvent)=\"<function>\""
+// /(\b)__TWIC_STATE_CHANGE_IMG__(\b)/gm => "\n| `stateChangeEvent` | An event emitter triggered each time the image loading state is updated. State can be `new`, `loading`, `loaded` or `error`.| [`EventEmitter<StateEvent>`](#state-event-type) | |"
```
Comment (Strift) [inline src/_documentation/angular.md:7]: Should be `done` instead of `loaded`

File: src/_documentation/angular.md [hunk 4]
```diff
@@ -323,6 +328,49 @@ Particularly useful when creating hero banner, you can specify the height of you
   <img alt="Edit TwicPics x Angular - Hero Image" src="https://codesandbox.io/static/img/play-codesandbox.svg">
 </a>
 
+<div id='life-cycle'/>
```
Comment (Strift) [inline src/_documentation/angular.md:331]: ID should be `lifecycle`

File: src/_documentation/angular.md [hunk 5]
```diff
@@ -323,6 +328,49 @@ Particularly useful when creating hero banner, you can specify the height of you
...
+
+### Lifecycle
+
+Binding to `stateChangeEvent` gives access to the loading state of your image or video.
+
+Here are the values the TwicComponent will emit ([see State Type definition](#state-type)) :
```
Comment (Strift) [inline src/_documentation/angular.md:337]: Component not `TwicComponent`. It leads to believe `TwicComponent` is a real thing.

File: src/_documentation/angular.md [hunk 6]
```diff
@@ -323,6 +328,49 @@ Particularly useful when creating hero banner, you can specify the height of you
...
+
+- `new` when `img` or `video` source is not yet loading
+- `loading` when `img` or `video` source is being loaded
+- `loaded` when `img` or `video` source has being loaded
+- `error` when an error occurred while loading an `img` or `video` source
+
+
+`<your-page-or-component>.html`
```
Comment (Strift) [inline src/_documentation/angular.md:345]: Remove this line but add atop the HTML below:

`<!-- component.html -->`

(I'm writing it here but this is valid for all other examples too)

File: src/_documentation/angular.md [hunk 7]
```diff
@@ -323,6 +328,49 @@ Particularly useful when creating hero banner, you can specify the height of you
...
+
+```html
+<TwicImg
+  src="path/to/your/image"
+  (stateChangeEvent)="handleStateChange($event)"
+></TwicImg>
+```
+
+`<your-page-or-component>.ts`
```
Comment (Strift) [inline src/_documentation/angular.md:354]: Remove this line but add atop the TS below:

`// component.ts``

(Again, I'm writing it here, but it applies to all other examples)

File: src/_documentation/angular.md [hunk 8]
```diff
@@ -392,6 +440,36 @@ export class YourComponent {
 <TwicImg src="path/to/your/image" [placeholder]="yourPlaceholderVariable"></TwicImg>
 ```
 
+<div id='state-type'/>
+
+### State Type
```
Comment (Strift) [inline src/_documentation/angular.md:445]: Title should be: "Loading State Values"

File: src/_documentation/angular.md [hunk 9]
```diff
@@ -392,6 +440,36 @@ export class YourComponent {
...
+
+Union type for all possible state values emitted through `@output` property `stateChangeEvent`.
+
+```ts
+type State = `error` | `loaded` | `loading` | `new`;
```
Comment (Strift) [inline src/_documentation/angular.md:450]: Can this type be imported by the end user?

File: src/_documentation/angular.md [hunk 10]
```diff
@@ -392,6 +440,36 @@ export class YourComponent {
...
+```
+
+- `new` when `img` or `video` source is not yet loading
+- `loading` when `img` or `video` source is being loaded
+- `loaded` when `img` or `video` source has being loaded
+- `error` when an error occurred while loading an `img` or `video` source
```
Comment (Strift) [inline src/_documentation/angular.md:456]: I'd write:

- `new`: when the `img` or `video` source has not started loading
- `loading`: when the `img` or `video` source is loading
- `done`: when the `img` or `video` source has successfully loaded
- `error`: when an error occurred while loading the `img` or `video` source

File: src/_documentation/angular.md [hunk 11]
```diff
@@ -392,6 +440,36 @@ export class YourComponent {
...
+
+
+<div id='state-event-type'/>
+
+### StateEvent Type
```
Comment (Strift) [inline src/_documentation/angular.md:461]: Title should be: State Change Event

File: src/_documentation/angular.md [hunk 12]
```diff
@@ -392,6 +440,36 @@ export class YourComponent {
...
+
+Data type emitted by `@output` property `stateChangeEvent`.
+
+```ts
+export type StateEvent = {
+  target: TwicImgComponent | TwicVideoComponent,
+  state: State
+};
```
Comment (Strift) [inline src/_documentation/angular.md:469]: Is this type importable by the end user?

File: src/_documentation/common/_componentsProps.md [hunk 13]
```diff
@@ -11,21 +11,21 @@ This component can be used in place of an `img` element.
 ```html
 <TwicImg
   src="<path>" (mandatory)
```
Comment (Strift) [inline src/_documentation/common/_componentsProps.md:13]: remove `(mandatory)`, it prevents users from copy-pasting code

File: src/_documentation/common/_componentsProps.md [hunk 14]
```diff
@@ -57,20 +57,20 @@ This component can be used in place of a `video` element.
 ```html
 <TwicVideo
   src="<path>" (mandatory)
```
Comment (Strift) [inline src/_documentation/common/_componentsProps.md:59]: Remove `(mandatory)`

File: CHANGELOG.md (hunk 1 above)
Comment (mbgspcii) [reply CHANGELOG.md:13]: - [ ]

Comment (mbgspcii) [reply CHANGELOG.md:13]: - [ ] Update changelog

File: src/_documentation/angular.md (hunk 2 above)
Comment (mbgspcii) [reply src/_documentation/angular.md:5]: - [ ] Change anchor for #life-cycle --> #lifecycle

File: src/_documentation/angular.md (hunk 3 above)
Comment (mbgspcii) [reply src/_documentation/angular.md:7]: - [ ] change `loaded` to `done`

File: src/_documentation/angular.md (hunk 4 above)
Comment (mbgspcii) [reply src/_documentation/angular.md:331]: - [ ] ID should be lifecycle

File: src/_documentation/angular.md (hunk 5 above)
Comment (mbgspcii) [reply src/_documentation/angular.md:337]: - [ ] Replace TwicComponent by Component

File: src/_documentation/angular.md (hunk 6 above)
Comment (mbgspcii) [reply src/_documentation/angular.md:345]: - [ ] Modify sample title

File: src/_documentation/angular.md (hunk 7 above)
Comment (mbgspcii) [reply src/_documentation/angular.md:354]: - [ ] modify sample

File: src/_documentation/angular.md (hunk 9 above)
Comment (mbgspcii) [reply src/_documentation/angular.md:450]: Yes. 
I did some tests in the typed languages. See screenshot
image

File: src/_documentation/angular.md (hunk 10 above)
Comment (mbgspcii) [reply src/_documentation/angular.md:456]: - [ ] change description

File: src/_documentation/angular.md (hunk 11 above)
Comment (mbgspcii) [reply src/_documentation/angular.md:461]: - [ ] State Change Event

File: src/_documentation/common/_componentsProps.md (hunk 13 above)
Comment (mbgspcii) [reply src/_documentation/common/_componentsProps.md:13]: - [ ] remove (mandatory)

File: src/_documentation/common/_componentsProps.md (hunk 14 above)
Comment (mbgspcii) [reply src/_documentation/common/_componentsProps.md:59]: - [ ] remove (mandatory)

**Pull request merged**